semistruct.go

```diff
@@ -0,0 +1,271 @@
+		p.Tag("attrs", Attrs()),
+		skipSpace,
+
+		// Parse the ending sentinel ">!"
+		EndSentinel(),
+	)
+
+	o.Node(func(m p.Match) (p.Match, error) {
+
+		// If we're parsing, we absolutely want a priority - if that
+		// returns an error then we should return an error; the other
+		// types can have sane defaults.
+		pr := p.GetTag(m, "priority").(int64)
+		tg := p.GetTag(m, "tags").([]string)
+		at := p.GetTag(m, "attrs").(map[string]string)
+
+		return &SemistructLog{pr, tg, at}, nil
+	})
+
+	return o
+}
+
+// PriorityInt parses the log level priority indicator; this can only
+// be an integer from 0 to 9.
+func PriorityInt() *p.Grammar {
+	o := p.Set("0-9")
+	o.Node(func(m p.Match) (p.Match, error) {
+		v, err := strconv.ParseInt(p.String(m), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		return v, nil
+	})
+
+	return o
+}
+
+// AlphaNum parses zero or more characters that are alpha-numeric with
+// an underscore.
+func AlphaNum() *p.Grammar {
+	return p.Mult(
+		0, 0, p.Set(`\w\-`),
+	)
+}
+
+// AlphaUpperNum parses zero or more characters that are uppercase
+// alphabetic, numeric, and underscore.
+func AlphaUpperNum() *p.Grammar {
+	return p.Mult(
+		0, 0, p.Set("A-Z0-9_"),
+	)
+}
+
+// AlphaNumSpecial parses zero or more characters that are
+// alpha-numeric, space characters, and special characters.
+func AlphaNumSpecial() *p.Grammar {
+	return p.Mult(
+		0, 0,
+		p.Set(`\w\-~ `+"`"+`\t!@#:;\$%^&\*\(\)\+=\?\\\/><,\.\{\}\[\]\|'`),
+	)
+}
+
+// Alpha parses zero or more characters that are alphabetic and
+// underscore.
+func Alpha() *p.Grammar {
+	return p.Mult(
+		0, 0, p.Set("a-zA-Z_"),
+	)
+}
+
+// QuotedStr parses an `AlphaNumSpecial` string wrapped in double
+// quotes.
+func QuotedStr() *p.Grammar {
+	return p.And(
+		p.Ignore(p.Lit(`"`)),
+		AlphaNumSpecial(),
+		p.Ignore(p.Lit(`"`)),
+	)
+}
+
+// OpenSentinel consumes the opening sentinel "!<" character sequence.
+func OpenSentinel() *p.Grammar {
+	return p.Ignore(
+		p.And(p.Lit("!"), p.Lit("<")),
+	)
+}
+
+// EndSentinel consumes the ending sentinel ">!" character sequence.
+func EndSentinel() *p.Grammar {
+	return p.Ignore(
+		p.And(p.Lit(">"), p.Lit("!")),
+	)
+}
+
+// SkipSpace Consumes whitespace.
+func SkipSpace() *p.Grammar {
+	return p.Ignore(
+		p.Mult(0, 0, p.Set(`\t `)),
+	)
+}
+
+// Tag parses zero or more tag atoms (tag1:tag2) separated by a colon.
+func Tag() *p.Grammar {
+	alphaNum := AlphaNum()
+
+	o := p.Optional(
+		p.And(
+			p.Tag("tag", alphaNum),
+			p.Mult(
+				0, 0, p.And(
+					p.Ignore(p.Lit(":")),
+					p.Tag("tag", alphaNum),
+				),
+			),
+		),
+	)
+
+	o.Node(func(m p.Match) (p.Match, error) {
+		var tagAccum []string
+		for _, op := range p.GetTags(m, "tag") {
+			tagAccum = append(tagAccum, p.String(op))
+		}
+		return tagAccum, nil
+	})
+	return o
+}
+
+// Tags parses a high-level set of colon separated tag atoms bracketed
+// by square brackets: [tag1:tag2].
+func Tags() *p.Grammar {
+	o := p.Optional(
+		p.And(
+			p.Ignore(p.Lit("[")),
```

Comment:
Because the `Lit` combinator is only matching a literal string, not a regex pattern.